[Utility] E2E Feature Path Template (#734)
pokt-network/pocket: utility/doc/E2E_FEATURE_PATH_TEMPLATE.md

# E2E Feature Path <!-- omit in toc -->

_IMPROVE(olshansky): Once we've completed the entire process at least once, we'll add links to each step._

- [Introduction \& Goals](#introduction--goals)
- [Developer Journey](#developer-journey)

> - [Feature Specification](#feature-specification)

Reviewer: ```suggestion
- [E2E Feature Specification](#e2e-feature-specification)
```

  - [Spot E2E Feature](#spot-e2e-feature)
  - [Spike E2E Feature](#spike-e2e-feature)
  - [Scope E2E Feature](#scope-e2e-feature)
    - [GitHub Ticket](#github-ticket)
    - [Origin Document](#origin-document)
- [E2E Feature Implementation](#e2e-feature-implementation)
  - [POC: Proof of Concept](#poc-proof-of-concept)
  - [MVP: Minimum Viable Product](#mvp-minimum-viable-product)
  - [PROD: Production](#prod-production)

## Introduction & Goals

> The [Pocket Network Specification](https://github.com/pokt-network/pocket-network-protocol/tree/main/utility) implementation is driven by various [milestones](https://github.com/pokt-network/pocket/milestones) and protocol/module/component specific tasks. Each feature cross the boundaries of business logic, data types, and interfaces for different components. Due to the complex nature of implementation, we've designed a more streamlined approach.

Reviewer: ```suggestion
The [Pocket Network Specification](https://github.com/pokt-network/pocket-network-protocol/tree/main/utility) implementation is driven by various [milestones](https://github.com/pokt-network/pocket/milestones) and protocol/module/component specific tasks. Each feature crosses the boundaries of business logic, data types, and interfaces for different components. Due to the complex nature of implementation, we've designed a streamlined approach which optimizes for supporting concurrent and iterative development.
```

> **The goal** of this document is to outline a well-defined process for adopting an end-to-end feature path approach. This makes each feature/task easier to scope, reason about, design, and implement.

Reviewer: ```suggestion
**The goal** of this document is to outline a well-defined process for adding an end-to-end feature path. This makes each feature/task easier to scope, reason about, design, and implement.
```

## Developer Journey

```mermaid
%%{init: { 'logLevel': 'debug', 'theme': 'base' } }%%
  timeline
    title E2E Feature Path Developer Journey

    section E2E Feature Specification
      Spot Feature:
        User research:
        Product management:
        Protocol intuition/experience:
        Addition/removal/selection of feature from the list
      Spike Feature:
        Research details:
        Identify pointers:
        Note dependencies:
        Find blockers
      Scope Feature:
        Define requirements:
        Document E2E implementation

    section E2E Feature Implementation
        POC:
          POC Spike:
          Explore:
          Hack:
          Have fun!
        MVP:
          Data Structures:
          Interfaces:
          Implementation:
          Unit Tests:
          E2E Tests:
          Documentation
        PROD:
          Identify workarounds & hacks:
          Document future work:
          Ideate!
```

> ## Feature Specification

Reviewer: ```suggestion
## E2E Feature Specification
```

### Spot E2E Feature

> Choose or add a feature from the Utility E2E feature list [here](./E2E_FEATURE_LIST.md).

Reviewer: Is it intentional that this file isn't present yet?

Author: Yea, that is up for review in #766 and should be merged in shortly as well.

> ### Spike E2E Feature

Reviewer: ```suggestion
### Spike Feature
```

Create a SPIKE GitHub issue, like [this](<[http](https://github.com/pokt-network/pocket/issue/TODO_LINK_TO_ISSUE_ONCE_WE_HAVE_EXAMPLE)>) to scope the feature. This ticket is responsible for creating the ticket that'll track the work.

> ### Scope E2E Feature

Reviewer: ```suggestion
### Scope Feature
```

Leverage the results from the SPIKE to create an implementation GitHub issue, like [this](<[http](https://github.com/pokt-network/pocket/issue/TODO_LINK_TO_ISSUE_ONCE_WE_HAVE_EXAMPLE)>) to track the actual implementation.

> #### GitHub Ticket

Reviewer: I find it difficult to make the distinction between an H4 and a bold text:

![image](https://github.com/pokt-network/pocket/assets/600733/cc6eaa10-6f3c-4e9c-9382-b0de9b99b0ce)

"Scope E2E Feature" is noticeably larger than the rest of the text, but "GitHubTicket" and "Objective" are indistinguishable to my eye.

As the majority of the details in the "E2E Feature Specification" section are in the H3/H4 levels of the hierarchy, I wonder if there's a convenient way to push the detail "up" one level? Perhaps "E2E Feature Specification" and "E2E Feature Implementation" could become H1s.

(FWIW, I also noticed that notion only supports H1-3, I would it's for similar UX reasons.)

Author: I added numbering. This way you know it's a section, it shows up in the ToC, and we don't need to overthink it too much.

Let me know if you'd like me to make any further changes

> Adhering to the GitHub template provided [here](https://github.com/pokt-network/pocket/blob/main/.github/ISSUE_TEMPLATE/issue.md), generate a [new issue](https://github.com/pokt-network/pocket/issues/new?assignees=&labels=&projects=&template=issue.md&title=%5BREPLACE+ME%5D+with+a+descriptive+title) encompassing the subsequent elements:

Reviewer: Is there any value in linking to the issue template as the template automatically populates the issue description when following the "new issue" link? I find it a bit distracting otherwise.

```suggestion
Open a [new issue](https://github.com/pokt-network/pocket/issues/new?assignees=&labels=&projects=&template=issue.md&title=%5BREPLACE+ME%5D+with+a+descriptive+title) and populate its description, respectively, with the following additional elements:
```

**Objective**: `Implement MVP E2E Feature Path <Letter>.<Number>: <Name>`

**Origin Document**: _Specify the details from the [Origin Document](#origin-document) below_

**Goals**:

- Complete the MVP implementation of the E2E Feature Path outlined in the objective
- Identify future tasks and test requirements to transition the feature to production

**Deliverables**:

**POC**:

> - [ ] A POC SPIKE to be closed out and split out into multiple PRs

Reviewer: nit + personal opinion: I would prefer "discarded" over "closed out" because I don't think a spike necessarily needs to be opened as a huge PR and closed ritualistically. If that's the process we find to be most useful, I think it still fits within the more general notion of "discarded". I would also prefer "refactored and/or restructured" over "split out" to explicitly call out the desire to prioritize improving quality at this stage to such an extent that it likely necessitates code refactoring as well as restructuring commits (contents & order). 

_I'm open to couter-suggestions :stuck_out_tongue_winking_eye:_

```suggestion
- [ ] A POC SPIKE to be discarded, refactored, and/or restructured into multiple PRs
```

> **MVP**:

Reviewer: nit: Wdyt about "MVC" for "minimum viable change"? I encountered this phrasing while reading through some of the [GitLab employee handbook (values > iteration)](https://handbook.gitlab.com/handbook/values/#minimal-viable-change-mvc) and I think it could be a useful distinction for us to describe "minimums" for sub-product scope tasks.

Author: LOVE IT!

- [ ] A PR that adds or modifies relevant structures and interfaces; such as [shared/core/types/proto](../../shared/core/types/proto), [shared/modules](../../shared/modules), etc
- [ ] A PR that materializes an MVP of the feature along with unit tests

> - [ ] A PR that introduces a new E2E tests with **one happy** and **one sad** path scenarios as described in the origin document (refer to [e2e/README.md](../../e2e/e2e/README.md)); this may require additions to the [cli](https://github.com/pokt-network/pocket/tree/main/app/client)

Reviewer: Are we concerned about developers going overboard on test cases or would you be open to something like:
```suggestion
- [ ] A PR that introduces a new E2E tests with **at least one happy** and **at least one sad** path scenarios as described in the origin document (refer to [e2e/README.md](../../e2e/e2e/README.md)); this may require additions to the [cli](https://github.com/pokt-network/pocket/tree/main/app/client)
```

> - [ ] A PR that introduces a new E2E tests with **one happy** and **one sad** path scenarios as described in the origin document (refer to [e2e/README.md](../../e2e/e2e/README.md)); this may require additions to the [cli](https://github.com/pokt-network/pocket/tree/main/app/client)

Reviewer: ```suggestion
- [ ] A PR that introduces a new E2E tests with **one happy** and **one sad** path scenarios as described in the origin document (refer to [e2e/README.md](../../e2e/README.md)); this may require additions to the [cli](https://github.com/pokt-network/pocket/tree/main/app/client)
```

> - [ ] A PR that introduces a new E2E tests with **one happy** and **one sad** path scenarios as described in the origin document (refer to [e2e/README.md](../../e2e/e2e/README.md)); this may require additions to the [cli](https://github.com/pokt-network/pocket/tree/main/app/client)

Reviewer: Broken link

```suggestion
- [ ] A PR that introduces a new E2E tests with **one happy** and **one sad** path scenarios as described in the origin document (refer to [e2e/README.md](../../e2e/README.md)); this may require additions to the [cli](https://github.com/pokt-network/pocket/tree/main/app/client)
```

- [ ] A PR that updates all pertinent documentation

**PROD**:

- [ ] One or more subsequent GitHub issues that track future work including, but not limited to:
  - Enhancing test coverage
  - Adding subsequent features
  - Patching hacks or workarounds
  - Enabling your [imagination](https://github.com/pokt-network/pocket/assets/1892194/6aff9004-8d3b-48e8-b6d5-9b67ac266e3d)!

> #### Origin Document

Reviewer: Is there a particular reason not to move this to come before the goals and deliverables, as it does in the description template? I would imagine the UX would be smoother with less scrolling.

Author: The idea was to show wha the ticket looks like with this comment

```
**Origin Document**: _Specify the details from the [Origin Document](#origin-document) below_
```

Then the "meat of it" (i.e. the origin document) is outlined in detail below. It felt simpler when I was writing it but happy to modify.

> **Purpose:** A single sentence that captures the intended purpose, behaviour and goal of the E2E feature

Reviewer: It's not clear if you intend for developers to copy/paste this line and write that sentence below or if it is a placeholder for the sentence it describes.

**Actors**: Check all of the protocol actors involved in the feature:

- [ ] Validator
- [ ] Application
- [ ] Servicer
- [ ] Fisherman

> - [ ] Portal

Reviewer: :100:

**Data Structures**:

- A list of the core types (protobufs, structs, etc) that will be used, added or modified in this feature
- Mention or link to specific files if applicable
- See [shared/core/types/proto](../../shared/core/types/proto) as a reference as they will most likely, but not necessarily, be part of that package
- _TIPS:_

  - _This will be non-exhaustive and will likely change during the POC or MVP stages_
  - You can find all other structs by running this command:

    ```bash
    grep -r "type .* struct" --exclude-dir="vendor" --exclude="*.gen.go" --exclude="*.pb.go" .
    ```

  - You can find all other protobufs by running this command:

    ```bash
    find . -name "*.proto" -not -path "./vendor/*"`

>     ```

Author: Done!

**Interfaces**:

- A list of the interface (go interfaces, placeholder functions, grpc, etc) that will be used, added or modified in this feature
- Mention or link to specific files if applicable
- See [shared/modules](../../shared/modules) as a reference as they will most likely, but not necessarily, be part of that package
- _TIPS:_

  - _This will be non-exhaustive and will likely change during the POC or MVP stages_
  - You can find all other structs by running this command:

    ```bash
    grep -r "type .* interface" --exclude-dir="vendor" --exclude="*.gen.go" --exclude="*.pb.go" .
    ```

  - You can find all other protobufs by running this command:

    ```bash
    find . -name "*.proto" -not -path "./vendor/*"`

>     ```

Reviewer: :sunglasses: (same as above)

**Diagram**:

> - _One or more mermaid diagrams that will visualize the E2E feature_

Reviewer: Is there a particular reason for this line to be italicized?

Author: Nope :)

- _TIPs:_
  - Use multiple diagrams if a single one ends up exceeding 7 or more core elements or steps
  - See if there’s anything in [pokt-network/pocket-network-protocol/tree/main/utility](../../utility/) or [pokt-network/pocket/tree/main/utility/doc](../../utility/doc) that you can use as a starting point

**User Stories as Tests**:

- Use natural language (long-form or bullet points) to define:
  - One (or more) HAPPY E2E path(s) from start to end with all the relevant details
  - One (or more) SAD E2E path(s) from start to end with all the relevant details

>   - **[IMPROVE] Guiding template**: A [User | Actor | Source | etc] [performs an action] [where | when | at] [some specific context or state is guaranteed] and [the expected result is...].

Reviewer: :100: Great starting point! :raised_hands:

  - [Spot Feature](#spot-feature)
  - [Spike Feature](#spike-feature)
  - [Scope Feature](#scope-feature)
    - [1. GitHub Ticket](#1-github-ticket)

>     - [2. Origin Document](#2-origin-document)

Reviewer: Great solution! :raised_hands: (thanks for indulging me :sweat_smile:)

  - Patching hacks or workarounds
  - Enabling your [imagination](https://github.com/pokt-network/pocket/assets/1892194/6aff9004-8d3b-48e8-b6d5-9b67ac266e3d)!

#### 2. Origin Document

> **Purpose:** [Replace this with a single sentence that captures the intended purpose, behaviour and goal of the E2E feature]

Reviewer: :100:

- Mention or link to specific files if applicable
- See [shared/core/types/proto](../../shared/core/types/proto) as a reference as they will most likely, but not necessarily, be part of that package
- _TIPS:_
  - _This will be non-exhaustive and will likely change during the POC or MVC stages_
  - You can find all other structs by running `make search_structs`

>   - You can find all other protobufs by running `make search_protos`

Reviewer: :sunglasses:
